includes/data-stores/class-wc-admin-reports-customers-data-store.php

```diff
@@ -0,0 +1,399 @@
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			$total_pages = (int) ceil( $db_records_count / $sql_query_params['per_page'] );
+			if ( $query_args['page'] < 1 || $query_args['page'] > $total_pages ) {
+				return $data;
+			}
+
+			$customer_data = $wpdb->get_results(
+				"SELECT
+						{$selections}
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+					GROUP BY
+						customer_id
+					ORDER BY
+						{$sql_query_params['order_by_clause']}
+					{$sql_query_params['limit']}
+					",
+				ARRAY_A
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			if ( null === $customer_data ) {
+				return $data;
+			}
+
+			$customer_data = array_map( array( $this, 'cast_numbers' ), $customer_data );
+			$data          = (object) array(
+				'data'    => $customer_data,
+				'total'   => $db_records_count,
+				'pages'   => $total_pages,
+				'page_no' => (int) $query_args['page'],
+			);
+
+			wp_cache_set( $cache_key, $data, $this->cache_group );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Add customer information to the lookup table when orders are created or modified.
+	 *
+	 * @param int $post_id Post ID.
+	 */
+	public static function sync_order( $post_id ) {
+		if ( 'shop_order' !== get_post_type( $post_id ) ) {
+			return;
+		}
+
+		$order = wc_get_order( $post_id );
+		if ( ! $order ) {
+			return;
+		}
+
+		$customer_id = $order->get_customer_id();
+		if ( 0 === $customer_id ) {
+			return;
+		}
+
+		self::update_registered_customer( $customer_id );
+	}
+
+	/**
+	 * Update the database with customer data.
+	 *
+	 * @param int $user_id WP User ID to update customer data for.
+	 * @return int|bool|null Number or rows modified or false on failure.
+	 */
+	public static function update_registered_customer( $user_id ) {
+		global $wpdb;
+
+		$customer    = new WC_Customer( $user_id );
+		$order_count = $customer->get_order_count( 'edit' );
+		$total_spend = $customer->get_total_spent( 'edit' );
+		$last_order  = $customer->get_last_order();
+		$last_active = $customer->get_meta( 'wc_last_active', true, 'edit' );
+
+		// TODO: try to preserve customer_id for existing user_id?
+		return $wpdb->replace(
+			$wpdb->prefix . self::TABLE_NAME,
+			array(
+				'user_id'          => $user_id,
+				'username'         => $customer->get_username( 'edit' ),
+				'first_name'       => $customer->get_first_name( 'edit' ),
+				'last_name'        => $customer->get_last_name( 'edit' ),
+				'email'            => $customer->get_email( 'edit' ),
+				'city'             => $customer->get_billing_city( 'edit' ),
+				'postcode'         => $customer->get_billing_postcode( 'edit' ),
+				'country'          => $customer->get_billing_country( 'edit' ),
+				'orders_count'     => $order_count,
+				'total_spend'      => $total_spend,
+				'avg_order_value'  => $order_count ? ( $total_spend / $order_count ) : 0,
```

Comment:
should we also verify that $total_spend is > 0? Perhaps a customer has an order, but the order had zero value ( sounds strange, but i have "customers" like this on a site )

Reply:
I don't think we need to. If `$total_spend` is `0`, then the division operation results in `0`. If `$order_count` is `0` and we don't prevent the division, we get PHP warnings.

Reply:
TIL you can divide zero in PHP. :trollface: